File: CS/WindowsApplication3/ViewModels/DocumentsViewModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DXSample.Common.ViewModels;
using DXSample.MVVM;

namespace DXSample.ViewModels {
    public class DocumentsViewModel : ISupportLogicalLayout {
        const string ViewLayoutName = "DocumentsViewModel";
        public DocumentsViewModel() {
            Modules = CreateModules();
        }
        public ModuleDescription[] Modules { get; private set; }
        protected IDocumentManagerService DocumentManagerService { get { return this.GetService<IDocumentManagerService>(); } }
        protected ILayoutSerializationService LayoutSerializationService { get { return this.GetService<ILayoutSerializationService>(); } }
        #region ISupportLogicalLayout
        public IEnumerable<object> LookupViewModels {
            get { return null; }
        }
        IDocumentManagerService ISupportLogicalLayout.DocumentManagerService {
            get { return DocumentManagerService; }
        }
        public bool CanSerialize {
            get { return true; }
        }
        #endregion
        public virtual void OnClosing(CancelEventArgs cancelEventArgs) {
            SaveLogicalLayout();
            if (LayoutSerializationService != null) {
                ViewModelLogicalLayoutHelper.PersistentViewsLayout[ViewLayoutName] = LayoutSerializationService.Serialize();
            }
            ViewModelLogicalLayoutHelper.SaveLayout();
        }
        public void SaveLogicalLayout() {
            ViewModelLogicalLayoutHelper.PersistentLogicalLayout = LogicalLayoutSerializationHelper.SerializeDocumentManagerService(this);
        }
        public void Show(ModuleDescription module) {
            ShowCore(module);
        }
        public IDocument ShowCore(ModuleDescription module) {
            if (module == null || DocumentManagerService == null)
                return null;
            IDocument document = DocumentManagerService.FindDocumentByIdOrCreate(module.DocumentType, x => CreateDocument(module));
            document.Show();
            return document;
        }
        IDocument CreateDocument(ModuleDescription module) {
            var document = DocumentManagerService.CreateDocument(module.DocumentType, null, this);
            document.Title = module.Title;
            document.DestroyOnClose = false;
            return document;
        }
        public virtual void OnLoaded() {
            RestoreLogicalLayout();
            string state = null;
            if (LayoutSerializationService != null && ViewModelLogicalLayoutHelper.PersistentViewsLayout.TryGetValue(ViewLayoutName, out state)) {
                LayoutSerializationService.Deserialize(state);
            }
        }
        public bool RestoreLogicalLayout() {
            if (string.IsNullOrWhiteSpace(ViewModelLogicalLayoutHelper.PersistentLogicalLayout))
                return false;
            this.RestoreDocumentManagerService(ViewModelLogicalLayoutHelper.PersistentLogicalLayout);
            return true;
        }
        protected ModuleDescription[] CreateModules() {
            return new ModuleDescription[] {
                new ModuleDescription("CustomerView", "Customers"),
                new ModuleDescription("OrderView", "Orders")
			};
        }
        
    }
}

File: VB/WindowsApplication3/ViewModels/DocumentsViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DXSample.Common.ViewModels
Imports DXSample.MVVM

Namespace DXSample.ViewModels
    Public Class DocumentsViewModel
        Implements ISupportLogicalLayout

        Private Const ViewLayoutName As String = "DocumentsViewModel"
        Public Sub New()
            Modules = CreateModules()
        End Sub
        Private privateModules As ModuleDescription()
        Public Property Modules() As ModuleDescription()
            Get
                Return privateModules
            End Get
            Private Set(ByVal value As ModuleDescription())
                privateModules = value
            End Set
        End Property
        Protected ReadOnly Property IDocumentManagerService() As IDocumentManagerService
            Get
                Return Me.GetService(Of IDocumentManagerService)()
            End Get
        End Property
        Protected ReadOnly Property LayoutSerializationService() As ILayoutSerializationService
            Get
                Return Me.GetService(Of ILayoutSerializationService)()
            End Get
        End Property
        #Region "ISupportLogicalLayout"
        Public ReadOnly Property LookupViewModels() As IEnumerable(Of Object) Implements ISupportLogicalLayout.LookupViewModels
            Get
                Return Nothing
            End Get
        End Property
        Private ReadOnly Property DocumentManagerService() As IDocumentManagerService Implements ISupportLogicalLayout.DocumentManagerService
            Get
                Return IDocumentManagerService
            End Get
        End Property
        Public ReadOnly Property CanSerialize() As Boolean Implements ISupportLogicalLayout.CanSerialize
            Get
                Return True
            End Get
        End Property
        #End Region
        Public Overridable Sub OnClosing(ByVal cancelEventArgs As CancelEventArgs)
            SaveLogicalLayout()
            If LayoutSerializationService IsNot Nothing Then
                ViewModelLogicalLayoutHelper.PersistentViewsLayout(ViewLayoutName) = LayoutSerializationService.Serialize()
            End If
            ViewModelLogicalLayoutHelper.SaveLayout()
        End Sub
        Public Sub SaveLogicalLayout()
            ViewModelLogicalLayoutHelper.PersistentLogicalLayout = LogicalLayoutSerializationHelper.SerializeDocumentManagerService(Me)
        End Sub
        Public Sub Show(ByVal [module] As ModuleDescription)
            ShowCore([module])
        End Sub
        Public Function ShowCore(ByVal [module] As ModuleDescription) As IDocument
            If [module] Is Nothing OrElse DocumentManagerService Is Nothing Then
                Return Nothing
            End If
            Dim document As IDocument = DocumentManagerService.FindDocumentByIdOrCreate([module].DocumentType, Function(x) CreateDocument([module]))
            document.Show()
            Return document
        End Function
        Private Function CreateDocument(ByVal [module] As ModuleDescription) As IDocument
            Dim document = DocumentManagerService.CreateDocument([module].DocumentType, Nothing, Me)
            document.Title = [module].Title
            document.DestroyOnClose = False
            Return document
        End Function
        Public Overridable Sub OnLoaded()
            RestoreLogicalLayout()
            Dim state As String = Nothing
            If LayoutSerializationService IsNot Nothing AndAlso ViewModelLogicalLayoutHelper.PersistentViewsLayout.TryGetValue(ViewLayoutName, state) Then
                LayoutSerializationService.Deserialize(state)
            End If
        End Sub
        Public Function RestoreLogicalLayout() As Boolean
            If String.IsNullOrWhiteSpace(ViewModelLogicalLayoutHelper.PersistentLogicalLayout) Then
                Return False
            End If
            Me.RestoreDocumentManagerService(ViewModelLogicalLayoutHelper.PersistentLogicalLayout)
            Return True
        End Function
        Protected Function CreateModules() As ModuleDescription()
            Return New ModuleDescription() { _
                New ModuleDescription("CustomerView", "Customers"), _
                New ModuleDescription("OrderView", "Orders") _
            }
        End Function

    End Class
End Namespace
